HeterogeneousCore/SonicCore/interface/SonicClientBase.h

```diff
@@ -0,0 +1,49 @@
+#ifndef HeterogeneousCore_SonicCore_SonicClientBase
+#define HeterogeneousCore_SonicCore_SonicClientBase
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <string>
+#include <chrono>
+#include <exception>
+
+class SonicClientBase {
+public:
+  //destructor
+  virtual ~SonicClientBase() {}
+
+  void setDebugName(const std::string& debugName) { debugName_ = debugName; }
+
+  //main operation
+  virtual void predict(edm::WaitingTaskWithArenaHolder holder) = 0;
+
+protected:
+  virtual void predictImpl() = 0;
+
+  void setStartTime() {
+    if (debugName_.empty())
+      return;
+    t0_ = std::chrono::high_resolution_clock::now();
+    setTime_ = true;
+  }
+
+  void finish(std::exception_ptr eptr = std::exception_ptr{}) {
```

Comment:
If I understood the logic correctly, one thing that I find confusing (and thus do not like much) is that the `Sync` and `PseudoAsync` clients will automatically call `finish()`, while the `Async` client delegates the user to do so.

Reply:
Yes, this is unfortunate but unavoidable. For a real-world example, see:
https://github.com/hls-fpga-machine-learning/SonicCMS/blob/v5.0.0/TensorRT/src/TRTClient.cc#L112

`finish()` has to be executed in the server's callback. There's no generic way to do it on the client side when using a truly asynchronous call.

Reply:
Yes, I understand.

What about making the user call `finish()` also for the Sync and PseudoAsync cases ?

This would eliminate the confusion (and the specialisation in the dummy test).

Reply:
That would be more consistent, but less user-friendly... to some extent I *want* to denote that true Async is a special case (some protocols don't support it, requires a callback, etc.). I'll think about it a bit more.

Reply:
I've implemented this change now, and added a note about it to the README.